pkg/codec/utils.go

```diff
@@ -11,3 +15,119 @@ func FitsInNBitsSigned(n int, bi *big.Int) bool {
 	}
 	return bi.BitLen() <= n-1
 }
+
+func BigIntHook(_, to reflect.Type, data any) (any, error) {
+	if to == reflect.TypeOf(&big.Int{}) {
+		bigInt := big.NewInt(0)
+
+		switch v := data.(type) {
+		case float64:
+			bigInt.SetInt64(int64(v))
+		case float32:
+			bigInt.SetInt64(int64(v))
+		case int:
+			bigInt.SetInt64(int64(v))
+		case int8:
+			bigInt.SetInt64(int64(v))
+		case int16:
+			bigInt.SetInt64(int64(v))
+		case int32:
+			bigInt.SetInt64(int64(v))
+		case int64:
+			bigInt.SetInt64(v)
+		case uint:
+			bigInt.SetUint64(uint64(v))
+		case uint8:
+			bigInt.SetUint64(uint64(v))
+		case uint16:
+			bigInt.SetUint64(uint64(v))
+		case uint32:
+			bigInt.SetUint64(uint64(v))
+		case uint64:
+			bigInt.SetUint64(v)
+		case string:
+			_, ok := bigInt.SetString(v, 10)
+			if !ok {
+				return nil, types.InvalidTypeError{}
+			}
+		default:
+			return data, nil
+		}
+
+		return bigInt, nil
+	}
+
+	if bi, ok := data.(*big.Int); ok {
+		switch to {
+		case reflect.TypeOf(0):
+			if FitsInNBitsSigned(strconv.IntSize, bi) {
+				return int(bi.Int64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(int8(0)):
+			if FitsInNBitsSigned(8, bi) {
+				return int8(bi.Int64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(int16(0)):
+			if FitsInNBitsSigned(16, bi) {
+				return int16(bi.Int64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(int32(0)):
+			if FitsInNBitsSigned(32, bi) {
+				return int32(bi.Int64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(int64(0)):
+			if FitsInNBitsSigned(64, bi) {
+				return bi.Int64(), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(uint(0)):
+			if bi.Sign() >= 0 && bi.BitLen() <= strconv.IntSize {
+				return uint(bi.Uint64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(uint8(0)):
+			if bi.Sign() >= 0 && bi.BitLen() <= 8 {
+				return uint8(bi.Uint64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(uint16(0)):
+			if bi.Sign() >= 0 && bi.BitLen() <= 16 {
+				return uint16(bi.Uint64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(uint32(0)):
+			if bi.Sign() >= 0 && bi.BitLen() <= 32 {
+				return uint32(bi.Uint64()), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(uint64(0)):
+			if bi.Sign() >= 0 && bi.BitLen() <= 64 {
+				return bi.Uint64(), nil
+			}
+			return nil, types.InvalidTypeError{}
+		case reflect.TypeOf(""):
+			return bi.String(), nil
+		default:
+			return data, nil
+		}
+	}
+
+	return data, nil
+}
+
+func SliceToArrayVerifySizeHook(from reflect.Type, to reflect.Type, data any) (any, error) {
+	// By default, if the array is bigger it'll still work. (ie []int{1, 2, 3} => [4]int{} works with 0 at the end
+	// [2]int{} would not work. This seems to lenient, but can be discussed.
```

Comment:
Yeah, I suspect we'll want to enforce that the length matches exactly, but I guess we should gather feedback from others about use cases. This is fine for now.